collector/osquery_collector.go

```diff
@@ -122,34 +132,38 @@ func (c *OsqueryCollector) Describe(ch chan<- *prometheus.Desc) {
 
 // Collect implements prometheus.Collector
 func (c *OsqueryCollector) Collect(ch chan<- prometheus.Metric) {
-	wg := sync.WaitGroup{}
-	wg.Add(len(c.collectors))
-	for _, col := range c.collectors {
-		go func(col singleQueryCollector) {
-			defer wg.Done()
-			result, err := c.runner.Run(col.Query())
-			if err != nil {
-				log.Errorf("failed to run query %s: %s", col.Query(), err)
-				c.success.WithLabelValues(col.String()).Set(0.0)
-				return
-			}
-			c.resultsets.WithLabelValues(col.String()).Set(float64(len(result.Items)))
-			err = update(col, result, ch)
-			if err != nil {
-				log.Warnf("metric %s errors on update: %s", col.String(), err)
-				c.success.WithLabelValues(col.String()).Set(0.0)
-				return
-			}
-			log.Debugf("metric %s took %s seconds to run", col.String(), result.Runtime)
-			c.queryDurations.WithLabelValues(col.String()).Observe(
-				result.Runtime.Seconds(),
-			)
-			c.success.WithLabelValues(col.String()).Set(1.0)
-		}(col)
+	c.throttle.Lock.Lock()
+	defer c.throttle.Lock.Unlock()
+	if time.Now().After(c.throttle.LastRun.Add(c.throttle.Interval)) {
```

Comment:
this throttles all metrics, even if just one metric is slow. These should be a throttle state per osquery (based on the same interval for convenience and simplicity of the config).

Reply:
Agree, was my initial thought but went with easier approach to speed up a bit. Will take another look.